products/firebase/api.yaml

```diff
@@ -100,3 +100,91 @@ objects:
         description: |
           The ID of the default GCP resource location for the Project. The location must be one of the available GCP
           resource locations.
+  - !ruby/object:Api::Resource
+    name: 'WebApp'
+    min_version: beta
+    base_url: projects/{{project}}/webApps
+    self_link: '{{name}}'
+    update_verb: :PATCH
```

Comment:
Updating the display name doesn't currently work because the patch endpoint doesn't return an operation. You will need to add the async block under the resource definition instead of just at the top level. This will let you specify the `actions` that the async will apply to. 

eg:
```
    async: !ruby/object:Api::OpAsync
      actions: ["create"]
      operation: !ruby/object:Api::OpAsync::Operation
        path: 'name'
```

Reply:
Can you also add an update test? This will help catch these kind of failures in the future.

Reply:
Hmmm unfortunately when I did this, I now get a compilation error because of an used "res" variable returned from "sendRequestWithTimeout". Checking 'templates/terraform/resource.erb' I see 
```
<%      if object.async.is_a? Api::OpAsync-%>
        res, err := sendRequestWithTimeout(config, "<%= key[:update_verb] -%>", <% if has_project || object.supports_indirect_user_project_override %>project<% else %>""<% end %>, url, obj, d.Timeout(schema.TimeoutUpdate)<%= object.error_retry_predicates ? ", " + object.error_retry_predicates.join(',') : "" -%>)
<%      else -%>
        _, err = sendRequestWithTimeout(config, "<%= key[:update_verb] -%>", <% if has_project || object.supports_indirect_user_project_override %>project<% else %>""<% end %>, url, obj, d.Timeout(schema.TimeoutUpdate)<%= object.error_retry_predicates ? ", " + object.error_retry_predicates.join(',') : "" -%>)
<%      end -%>
```

Shouldn't the if statement say `object.async.is_a? Api::OpAsync && object.async&.allow?('update')` so that it only renders the "res" if its both async and update is an allowed action?

Reply:
I think this might be a previously unknown bug. I can find no examples of another resource that only supports 'create' for "OpAsyc" except ones that completely exclude generation.

Also, should I modify the api.yaml and move all the async declarations to the resource level? Or should I just an override in terraform.yaml for the WebApp resource?

Reply:
Yeah that's probably a bug. I'll either add a fix to this PR or separately so you can rebase.

Reply:
I'll add it. It's cool :). I just gotta get the test cases done for update and data source and I should be good.

Reply:
go for it!

Reply:
Unfortunately I've run in to a snag that I'm not sure what to do about.  I'll check in my changes so you can see but from a high level, even though the Firebase Project Location has been finalized and the web app's create operation has completed, when an API call is made to the "getConfig" endpoint, only the appId and projectId are returned. However a few seconds later, the same API call returns all the values completely filled in.

 I'm writing the 'update' test now and I'm verifying that the calculated configuration matches the one returned by the Go Firebase client.  I have a breakpoint set in test check func and I can see the google_project_location is finalized and the web app operation is complete (as i can see the data fields), but the response from the "getClient" api call is lacking all attributes. (see `testAccFirebaseWebAppConfig`) so the test fails.

This implies that the Operation is somehow completing before it's really finished.

Thoughts?

Reply:
I see there being 2 possible approaches:
There is a PollAsync class that it meant to allow resources to poll against the resource endpoint until the state transitions to a "done" state. This has been used for pub sub because it has been similarly slow to actually update. 
But my preference is probably to split the `getConfig` endpoint out into a separate datasource. Right now it is being shimmed in through the separate call in the decoder so it will be tough to leverage any of the first party Magic Modules features such as poll async.

Reply:
That was my original intent. I created a WebAppConfig resource, but if I do that, there was no way to get it excluded from the provider.go. I tried "readonly". "exclude" excludes it completely but still puts it in.  

I'm starting to get past my league on this. I'm happy to let you take over any part.

Reply:
That sounds fine. I'll probably just create the data source without a resource at all. If there is an endpoint that is truly read only there there's no need to manage it via a resource. I should be able to pick this up tomorrow or next week though!